### PR TITLE
[core]Migrate Convert operator to new API

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Migrate `convert` operator to new API.
 - Refactor `Convert` to not use legacy function and reduce binary size.
 - Unify conversion for non low precision types.

### Tickets:
 - [CVS-118489](https://jira.devtools.intel.com/browse/CVS-118489)

### CODE REVIEW

File: src/core/src/op/convert.cpp [hunk 1]
```diff
@@ -2,248 +2,198 @@
...
-        ov::reference::convert(arg->get_data_ptr<INPUT_ET>(), out->get_data_ptr<OUTPUT_ET>(), element_count);
-    }
-    return true;
+constexpr bool is_lp_type(const element::Type_t et) {
+    return (et == element::i4) || (et == element::u1) || (et == element::u4) || (et == element::nf4);
 }
 
-#define TYPE_OUT_CASE(a, ...)                                     \
-    case element::Type_t::a: {                                    \
-        OV_OP_SCOPE(OV_PP_CAT3(evaluate_covert_out, _, a));       \
```
Comment (vurusovs) [inline src/core/src/op/convert.cpp:70]: `OV_OP_SCOPE` not ported, so conditional compilation won't work as previously. @praasz is it possible to support same behaviour?

**Review by vurusovs — CHANGES_REQUESTED** [review]

File: src/core/src/op/convert.cpp [hunk 2]
```diff
@@ -2,248 +2,192 @@
...
+#define CONVERT_ET_LIST boolean, bf16, f16, f32, f64, i4, i8, i16, i32, i64, u1, u4, u8, u16, u32, u64, nf4
+
+struct Evaluate : public element::NoAction<bool> {
+    using element::NoAction<bool>::visit;
+    template <element::Type_t ET, class TO = fundamental_type_for<ET>>
+    static result_type visit(const Tensor& arg, Tensor& out, const size_t count) {
+        using namespace ov::element;
+        return IfTypeOf<CONVERT_ET_LIST>::apply<EvalByOutputType<is_lp_type(ET)>>(
+            out.get_element_type(),
+            reinterpret_cast<const TO*>(arg.data()),
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/src/op/convert.cpp:31]: Yes it is. Is just wrong name will change it.

File: src/core/src/op/convert.cpp (hunk 1 above)
Comment (praasz) [reply src/core/src/op/convert.cpp:70]: It looks like using `IfTypeOf`` missing this feature. 
I've created the ticket [CVS-124984](https://jira.devtools.intel.com/browse/CVS-124984) to resolve this in next step.
It has to be implemented in generic way for *IfTypeOf* to enable conditional compilation feature.

**Comment by mlukasze** [comment]
can we merge it?

**Review by t-jankowski — APPROVED** [review]

**Review by vurusovs — APPROVED** [review]

File: src/core/reference/include/openvino/reference/convert.hpp [hunk 3]
```diff
@@ -94,13 +95,21 @@ void lp_convert(const TI* arg, TO* out, size_t count, element::Type_t src_type,
...
 }
+
+template <typename TI, typename TO>
+typename std::enable_if<!std::is_same<TO, char>::value, TO>::type convert(const TI v) {
+    return static_cast<TO>(v);
+}
+
+template <typename TI, typename TO>
+typename std::enable_if<std::is_same<TO, char>::value, TO>::type convert(const TI v) {
+    return static_cast<bool>(v);
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/reference/include/openvino/reference/convert.hpp:106]: Done.

**Pull request merged**